databricks/koalas/indexing.py

```diff
@@ -400,16 +395,56 @@ def _select_rows(self, rows_sel):
                 # If slice is None - select everything, so nothing to do
                 return None, None
             elif len(self._internal.index_columns) == 1:
+                sdf = self._kdf_or_kser._internal.sdf
+                index = self._kdf_or_kser.index
+                index_column = index.to_series()
+                index_data_type = index_column.spark_type
                 start = rows_sel.start
                 stop = rows_sel.stop
+                start_order_column = sdf[NATURAL_ORDER_COLUMN_NAME]
+                stop_order_column = sdf[NATURAL_ORDER_COLUMN_NAME]
+
+                # get natural order from '__natural_order__' from start to stop
+                # to keep natural order.
+                start_and_stop = (
+                    sdf.select(index_column._scol, NATURAL_ORDER_COLUMN_NAME)
+                       .where((index_column._scol == start) | (index_column._scol == stop))
+                       .collect())
+
+                start = [row[1] for row in start_and_stop if row[0] == start]
+                start = start[0] if len(start) > 0 else None
+
+                stop = [row[1] for row in start_and_stop if row[0] == stop]
+                stop = stop[-1] if len(stop) > 0 else None
+
+                # if index order is not monotonic increasing or decreasing
+                # and specified values don't exist in index, raise KeyError
+                if start is None and rows_sel.start is not None:
+                    if not (index.is_monotonic_increasing or index.is_monotonic_decreasing):
+                        raise KeyError(rows_sel.start)
+                    else:
+                        start = rows_sel.start
+                        start_order_column = index_column._scol
+                if stop is None and rows_sel.stop is not None:
+                    if not (index.is_monotonic_increasing or index.is_monotonic_decreasing):
+                        raise KeyError(rows_sel.stop)
+                    else:
+                        stop = rows_sel.stop
+                        stop_order_column = index_column._scol
+
+                # if start and stop are same, just get all start(or stop) values
+                if start == stop:
+                    return index_column._scol == F.lit(rows_sel.start).cast(index_data_type), None
+
+                # we don't use StringType since we're using `__natural_order__` for comparing
+                if isinstance(index_data_type, StringType):
+                    index_data_type = LongType()
```

Comment:
I think this logic itself is incorrect. `start_order_column` can be either `index_column._scol` or `__natual_column__`. So, when the type is string, it can be either a string type or long type.

```python
>>> pd.DataFrame([1, 2, 3], index=['a', 'b', 'd']).loc['a':'z']
   0
a  1
b  2
d  3
```

```python
>>> ks.DataFrame([1, 2, 3], index=['a', 'b', 'd']).loc['a':'z']
Empty DataFrame
Columns: [0]
Index: []
```

Reply:
I think you should:


```python
                # Assume we use the natural order by default.
                start_order_column_type = LongType()
                stop_order_column_type = LongType()

                # if index order is not monotonic increasing or decreasing
                # and specified values don't exist in index, raise KeyError
                if start is None and rows_sel.start is not None:
                    if not (index.is_monotonic_decreasing or index.is_monotonic_increasing):
                        raise KeyError(rows_sel.start)
                    else:
                        start = rows_sel.start
                        start_order_column = index_column._scol
                        start_order_column_type = index_data_type
                if stop is None and rows_sel.stop is not None:
                    if not (index.is_monotonic_decreasing or index.is_monotonic_increasing):
                        raise KeyError(rows_sel.stop)
                    else:
                        stop = rows_sel.stop
                        stop_order_column = index_column._scol
                        stop_order_column_type = index_data_type

                # if start and stop are same, just get all start(or stop) values
                if start == stop:
                    return index_column._scol == F.lit(rows_sel.start).cast(index_data_type), None

                cond = []
                if start is not None:
                    cond.append(start_order_column >= F.lit(start).cast(start_order_column_type))
                if stop is not None:
                    cond.append(stop_order_column <= F.lit(stop).cast(stop_order_column_type))
```